README
NAME
    Types::URI - type constraints and coercions for URIs

SYNOPSIS
       package FroobleDocument;
   
       use Moose;
       use Types::URI -all;
   
       has source => (
          is      => 'ro',
          isa     => Uri,
          coerce  => 1,
       );

DESCRIPTION
    Types::URI is a type constraint library suitable for use with Moo/Moose
    attributes, Kavorka sub signatures, and so forth.

  Types
    This module provides some type constraints broadly compatible with those
    provided by MooseX::Types::URI, plus a couple of extra type constraints.

    `Uri`
        A class type for URI/URI::WithBase. Coercions from:

        from `Uuid`
            Coerces to a URI in the `urn:uuid:` schema. (See Types::UUID.)

        from `Str`
            Uses "new" in URI.

        from `Path`
            Uses "new" in URI::file. (See Types::Path::Tiny.)

        from `ScalarRef`
            Uses "new" in URI::data.

        from `HashRef`
            Coerces using URI::FromHash.

        from RDF::Trine::Node::Resource, RDF::Trine::Namespace, XML::Namespace
            Uses "new" in URI.

    `FileUri`
        A subtype of `Uri` covering URI::file. Coercions from:

        from `Str`
            Uses "new" in URI::file.

        from `Path`
            Uses "new" in URI::file. (See Types::Path::Tiny.)

        from `HashRef`
            Coerces using URI::FromHash.

        from RDF::Trine::Node::Resource, RDF::Trine::Namespace, XML::Namespace
            Uses "new" in URI.

    `DataUri`
        A subtype of `Uri` covering URI::data. Coercions from:

        from `Str`
            Uses "new" in URI::data.

        from `ScalarRef`
            Uses "new" in URI::data.

        from `HashRef`
            Coerces using URI::FromHash.

        from RDF::Trine::Node::Resource, RDF::Trine::Namespace, XML::Namespace
            Uses "new" in URI.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Types-URI>.

SEE ALSO
    MooseX::Types::URI, Type::Tiny::Manual, URI, URI::file, URI::data,
    URI::FromHash, RDF::Trine::Node::Resource.

    Types::UUID, Types::Path::Tiny, Types::Standard.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Types-URI
=========

Created:      2014-03-24
Home page:    <https://metacpan.org/release/Types-URI>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Types-URI>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.005	2014-09-27

 [ Packaging ]
 - Bump dependency on Type::Library to version 1.000000.

0.004	2014-09-27

 [ Packaging ]
 - Add dependency on Perl 5.8.

 [ Other ]
 - Added: Add a new Iri type, like Uri but using GWILLIAMS' IRI class.
   There are coercions between Uri and Iri.
 - Coercions are now made immutable.

0.003	2014-05-20

 - Added: Coercion from RDF::Trine::Namespace.
 - Added: Coercion from XML::Namespace.

0.002	2014-05-20

 [ Packaging ]
 - Repo metadata should not have been commented out.

 [ Other ]
 - Added: Coercion from RDF::Trine::Node::Resource.
 - Coercion from HashRef now also applies to DataUri and FileUri.

0.001	2014-04-13	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "type constraints and coercions for URIs",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.023, CPAN::Meta::Converter version 2.140640",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Types-URI",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { requires => { "Test::Warnings" => 0 } },
                        runtime   => {
                                       requires => {
                                         "perl" => 5.008,
                                         "Type::Library" => "1.000000",
                                         "Types::Path::Tiny" => 0,
                                         "Types::Standard" => 0,
                                         "Types::UUID" => 0,
                                         "URI" => 0,
                                         "URI::FromHash" => 0,
                                       },
                                     },
                        test      => {
                                       recommends => { IRI => 0.004, Moose => "2.0000" },
                                       requires   => { "Test::More" => 0.96, "Test::Requires" => 0 },
                                       suggests   => { "RDF::Trine::Node::Resource" => 0 },
                                     },
                      },
  "provides"       => {
                        "Types::URI" => { file => "lib/Types/URI.pm", version => 0.005 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Types-URI" },
                        homepage     => "https://metacpan.org/release/Types-URI",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-types-uri.git",
                                          web  => "https://github.com/tobyink/p5-types-uri",
                                        },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/Types-URI/project",
                      },
  "version"        => 0.005,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

